backend/updater.js

```diff
@@ -0,0 +1,250 @@
+
+      for (const classHash of user.watchingClasses) {
+        if (!classHashToUsers[classHash]) {
+          classHashToUsers[classHash] = [];
+        }
+
+        classHashToUsers[classHash].push(user.facebookMessengerId);
+      }
+
+      for (const sectionHash of user.watchingSections) {
+        if (!sectionHashToUsers[sectionHash]) {
+          sectionHashToUsers[sectionHash] = [];
+        }
+
+        sectionHashToUsers[sectionHash].push(user.facebookMessengerId);
+      }
+    }
+
+
+    const sectionHashMap = {};
+    const sections = [];
+
+    for (const sectionHash of sectionHashes) {
+      const aClass = this.dataLib.getSectionServerDataFromHash(sectionHash);
+
+      sections.push(aClass);
+      sectionHashMap[sectionHash] = aClass;
+    }
+
+
+    // Get the data for these hashes
+    const classes = [];
+    for (const classHash of classHashes) {
+      const aClass = this.dataLib.getClassServerDataFromHash(classHash);
+
+      classes.push(aClass);
+
+      for (const crn of aClass.crns) {
+        const sectionHash = Keys.create({
+          host: aClass.host,
+          termId: aClass.termId,
+          subject: aClass.subject,
+          classUid: aClass.classUid,
+          crn: crn,
+        }).getHash();
+
+        // Remove this one from the hash map
+        sectionHashMap[sectionHash] = false;
+      }
+    }
+
+
+    // Find the sections that are still around
+    for (const sectionHash of Object.keys(sectionHashMap)) {
+      // If it was set to false, ignore it
+      if (!sectionHashMap[sectionHash]) {
+        continue;
+      }
+
+
+      macros.error('Section', sectionHash, "is being watched but it's class is not being watched?", sectionHashMap);
+    }
+
+    let allParsersOutput = [];
+
+    // Scrape the latest data
+    for (const aClass of classes) {
+      const latestData = await ellucianCatalogParser.main(aClass.prettyUrl);
+
+      allParsersOutput = allParsersOutput.concat(latestData);
+    }
+
+    const rootNode = {
+      type: 'ignore',
+      deps: allParsersOutput,
+      value: {},
+    };
+
+
+    // Because ellucianCatalogParser returns a list of classes, instead of a singular class, we need to run it on all of them
+    const output = await classesScrapers.runProcessors(rootNode);
+
+
+    // Keep track of which messages to send which users. 
+    // The key is the facebookMessengerId and the value is a list of messages. 
+    let userToMessageMap = {};
+
+    for (const aNewClass of output.classes) {
+      const hash = Keys.create(aNewClass).getHash();
+
+      const oldClass = this.dataLib.getClassServerDataFromHash(hash);
+
+      // Count how many sections are present in the new but not in the old.
+      let count = 0;
+      for (const crn of aNewClass.crns) {
+        if (!oldClass.crns.includes(crn)) {
+          count++;
+        }
+      }
+
+      let message = '';
+      let classCode = `${aNewClass.subject} ${aNewClass.classId}`
```

Comment:
Is that space intentional? Just checking because it's a common mistake for me that has caused problems down the line.

Reply:
It depends where we are using the subject + classId combination tbh. The search supports with or without a space, so for a bunch of the hyperlinks I have been omitting the space so the link looks slightly cleaner. We could also add a plus between the subject and the classId and that would work too. Either one works for me really.

If we are showing this subject + classId to the user in text, I tend to add a space just so it is more readable. There are a bunch of places where either would work.